From b5cbb127036e13d26999a6ae19ce07823deb425d Mon Sep 17 00:00:00 2001
From: Rahul Tanwar <rtanwar@maxlinear.com>
Date: Tue, 22 Jun 2021 14:34:59 +0800
Subject: [PATCH] PONRTSYS-8955: Fix 'setting watchdog timeout on tep fails'

Fix below two bugs in TEP watchdog driver on PRX platform:
1. Setting watchdog timeout fails as reported in this JIRA.
2. Also found another bug that watchdog expiry fails to reset the system.
   Fix it by enabling GPTC2 Timer3 interrupt as well which is actually tied
   to WDT timer reset feature.
---
 drivers/watchdog/tep-wdt.c | 38 ++++++++++++++++++++++----------------
 1 file changed, 22 insertions(+), 16 deletions(-)

diff --git a/drivers/watchdog/tep-wdt.c b/drivers/watchdog/tep-wdt.c
index da0868d186cf..754a4ec3e5ec 100644
--- a/drivers/watchdog/tep-wdt.c
+++ b/drivers/watchdog/tep-wdt.c
@@ -83,11 +83,8 @@ static irqreturn_t tep_wdt_irq_handler(int irqno, void *param)
 	val |= RCU_GPHY_RESET;
 	regmap_write(priv->rcu, RST_REQ, val);
 
-	/* IRQ Mask */
-	val = GPTC_R32(priv->gptc, GPTC_IRNCR);
-	val &= ~BIT(TIMER_2 * 2);
-	GPTC_W32(priv->gptc, GPTC_IRNCR, val);
-
+	/* IRQ clear */
+	GPTC_W32(priv->gptc, GPTC_IRNCR, BIT(TIMER_2 * 2));
 	return IRQ_HANDLED;
 }
 
@@ -95,10 +92,17 @@ static int tep_wdt_start(struct watchdog_device *wdd)
 {
 	struct tep_wdt *priv = watchdog_get_drvdata(wdd);
 	u32 idx, tm = wdd->timeout;
+	u64 tm_out;
+
+	tm_out = ((u64)(priv->clk) * (u64)tm);
+	if ( tm_out <= (u64)U32_MAX)
+		tm = (u32)tm_out;
+	else
+		tm = U32_MAX;
 
 	for (idx = TIMER_2; idx <= TIMER_3; idx++) {
 		GPTC_W32(priv->gptc, GPTC_CON(idx), CON_ONESHOT | CON_EXT);
-		GPTC_W32(priv->gptc, GPTC_RLD(idx), ((tm == TIMER_2) ? (tm / 2) : tm));
+		GPTC_W32(priv->gptc, GPTC_RLD(idx), ((idx == TIMER_2) ? (tm / 2) : tm));
 		GPTC_W32(priv->gptc, GPTC_RUN(idx), RUN_EN | RUN_RELOAD);
 	}
 	return 0;
@@ -125,18 +129,10 @@ static int tep_wdt_ping(struct watchdog_device *wdd)
 
 static int tep_wdt_set_timeout(struct watchdog_device *wdd, u32 timeout)
 {
-	struct tep_wdt *priv = watchdog_get_drvdata(wdd);
-	u64 tm_out;
-
 	if (!capable(CAP_SYS_ADMIN))
 		return -EPERM;
 
-	tm_out = (u64)timeout * priv->clk;
-	if (tm_out > U32_MAX) {
-		tm_out = U32_MAX;
-		dev_warn(wdd->parent, "%s: set timeout overflow\n", __func__);
-	}
-	wdd->timeout = (u32)tm_out;
+	wdd->timeout = timeout;
 	return 0;
 }
 
@@ -252,8 +248,10 @@ static int tep_watchdog_probe(struct platform_device *pdev)
 	/* Enable WDT reset to RCU for VPEx */
 	regmap_write(priv->rcu, RCU_TEP_WDT_RST, WDT_SCPU_RST_EN);
 
-	/* GPTC2 timer2A */
+	/* Enable GPTC2 */
 	GPTC_W32(priv->gptc, GPTC_CLC, CLC_SUSPEND | CLC_RMC);
+
+	/* Enable GPTC2 Timer2 interrupt as prewarn timer interrupt */
 	GPTC_W32(priv->gptc, GPTC_IRNEN,
 		 ((~BIT(TIMER_2 * 2)) & GPTC_R32(priv->gptc, GPTC_IRNEN)));
 	GPTC_W32(priv->gptc, GPTC_IRNCR,
@@ -261,6 +259,14 @@ static int tep_watchdog_probe(struct platform_device *pdev)
 	GPTC_W32(priv->gptc, GPTC_IRNEN,
 		 ((BIT(TIMER_2 * 2)) | GPTC_R32(priv->gptc, GPTC_IRNEN)));
 
+	/* Enable GPTC2 Timer3 interrupt as watchdog reset interrupt */
+	GPTC_W32(priv->gptc, GPTC_IRNEN,
+		 ((~BIT(TIMER_3 * 2)) & GPTC_R32(priv->gptc, GPTC_IRNEN)));
+	GPTC_W32(priv->gptc, GPTC_IRNCR,
+		 ((BIT(TIMER_3 * 2)) | GPTC_R32(priv->gptc, GPTC_IRNCR)));
+	GPTC_W32(priv->gptc, GPTC_IRNEN,
+		 ((BIT(TIMER_3 * 2)) | GPTC_R32(priv->gptc, GPTC_IRNEN)));
+
 	return 0;
 }
 
